File: erlang/sum-of-multiples/sum_of_multiples.erl
-module(sum_of_multiples).
-export([sumOfMultiplesDefault/1, sumOfMultiples/2]).

sumOfMultiplesDefault(N) ->
  sumOfMultiples([3, 5], N).

sumOfMultiples(MultiplesOf, N) ->
  Multiples = [X || X <- lists:seq(0, N - 1),
                    lists:any(fun(E) -> X rem E =:= 0 end, MultiplesOf)],
  lists:foldl(fun(X, Sum) -> X + Sum end, 0, Multiples).

File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    Enum.sum(multiples(limit, factors))
  end

  defp multiples(limit, factors) do
    for x <- 1..(limit - 1), multiple?(x, factors), do: x
  end

  defp multiple?(n, factors) do
    Enum.any?(factors, &(rem(n, &1) == 0))
  end
end
